src/GafferScene/IECoreGLPreview/Renderer.cpp

```diff
@@ -161,24 +162,30 @@ class OpenGLAttributes : public IECoreScenePreview::Renderer::AttributesInterfac
 			IECoreGL::ConstStatePtr lightVisualisationState;
 			m_lightVisualisation = LightVisualiser::allVisualisations( attributes, lightVisualisationState );
 
-			if( m_lightVisualisation && m_visualisation )
+			IECoreGL::ConstStatePtr lightFilterVisualisationState;
+			m_lightFilterVisualisation = LightFilterVisualiser::allVisualisations( attributes, lightFilterVisualisationState );
+
+			if( m_lightFilterVisualisation )
 			{
-				// Light filter visualisers are in `m_visualisation` and light visualisers are in
-				// `m_lightVisualisation`. Combine them both into `m_lightVisualisation` so that
-				// filters attached to light locations are drawn as expected.
-				/// \todo We don't really want light filters to be in `m_visualisation` because then
-				/// they can get inherited onto locations below the light and drawn in duplicate.
-				/// Move `GafferSceneUI::LightFilterVisualiser` to `IECoreGL::LightFilterVisualiser` so
-				/// we can separate filter visualisations out, and apply them only to lights and
-				/// light filters.
-				IECoreGL::GroupPtr group = new IECoreGL::Group;
-				// `const_pointer_cast` ok because group becomes const on assignment to `m_lightVisualisation`
-				group->addChild( boost::const_pointer_cast<IECoreGL::Renderable>( m_visualisation ) );
-				group->addChild( boost::const_pointer_cast<IECoreGL::Renderable>( m_lightVisualisation ) );
-				m_lightVisualisation = group;
+				if( m_lightVisualisation )
+				{
+					// Light filter visualisers are in `m_lightFilterVisualisation` and light visualisers are in
+					// `m_lightVisualisation`. Combine them both into `m_lightVisualisation` so that
+					// filters attached to light locations are drawn as expected.
+					IECoreGL::GroupPtr group = new IECoreGL::Group;
+					// `const_pointer_cast` ok because group becomes const on assignment to `m_lightVisualisation`
+					group->addChild( boost::const_pointer_cast<IECoreGL::Renderable>( m_lightFilterVisualisation ) );
+					group->addChild( boost::const_pointer_cast<IECoreGL::Renderable>( m_lightVisualisation ) );
+					m_lightVisualisation = group;
```

Comment:
@johnhaddon Was this where you were thinking to merge the filter visualisations for lights? Feels slightly off being here, as they're duplicated, but it's the only place that holds a reference to the visualisations. `OpenGLLight` currently just bridges raw pointers, so grouping there there would need to keep the group alive there instead?

Reply:
Yep, this is exactly what I was expecting. OpenGLAttributes is responsible for holding visualisations of attributes, and light filters are attributes. We could monkey about with the `OpenGLObject::render()/visualisation()` stuff to allow multiple visualisations for lights, but I don't see any benefit over just grouping up front.